src/arch/x86_64/kernel/apic.rs

```diff
@@ -259,6 +315,145 @@ fn detect_from_acpi() -> Result<PhysAddr, ()> {
 	Ok(PhysAddr(madt_header.local_apic_address.into()))
 }
 
+/// Helper function to search Floating Pointer Structure of the Multiprocessing Specification
+fn search_mp_floating(start: PhysAddr, end: PhysAddr) -> Result<&'static ApicMP, ()> {
+	let virtual_address = virtualmem::allocate(BasePageSize::SIZE).map_err(|_| ())?;
+
+	for current_address in (start.as_usize()..end.as_usize()).step_by(BasePageSize::SIZE) {
+		let mut flags = PageTableEntryFlags::empty();
+		flags.normal().writable();
+		paging::map::<BasePageSize>(
+			virtual_address,
+			PhysAddr::from(align_down!(current_address, BasePageSize::SIZE)),
+			1,
+			flags,
+		);
+
+		for i in 0..BasePageSize::SIZE / 4 {
+			let mut tmp: *const u32 = virtual_address.as_ptr();
+			tmp = unsafe { tmp.offset(i.try_into().unwrap()) };
+			let apic_mp: &ApicMP = unsafe { &(*(tmp as *const ApicMP)) };
+			if apic_mp.signature == MP_FLT_SIGNATURE
+				&& !(apic_mp.version > 4 || apic_mp.features[0] != 0)
+			{
+				return Ok(apic_mp);
+			}
+		}
+	}
+
+	// frees obsolete virtual memory region for MMIO devices
+	virtualmem::deallocate(virtual_address, BasePageSize::SIZE);
+
+	Err(())
+}
+
+/// Helper function to detect APIC by the Multiprocessor Specification
+fn detect_from_mp() -> Result<PhysAddr, ()> {
+	let mp_float = if let Ok(mpf) = search_mp_floating(PhysAddr(0x9F000u64), PhysAddr(0xA0000u64)) {
+		Ok(mpf)
+	} else if let Ok(mpf) = search_mp_floating(PhysAddr(0xF0000u64), PhysAddr(0x100000u64)) {
+		Ok(mpf)
+	} else {
+		Err(())
+	}?;
+
+	info!("Found MP config at 0x{:x}", mp_float.mp_config);
```

Comment:
Instead of manually reading the unaligned field with `addr_of!` and `read_unaligned`, there is a more ergonomic solution for formatting:

Adding curly braces means that a copy of the field is made, stored in a (properly aligned) temporary, and a reference to that temporary is being formatted.

```suggestion
	info!("Found MP config at {:#x}", { mp_float.mp_config });
```